element-connector/src/main/java/org/eclipse/californium/elements/tcp/NettyContextUtils.java

```diff
@@ -0,0 +1,171 @@
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial implementation. 
+ *                                      add support for correlation context and principal
+ ******************************************************************************/
+package org.eclipse.californium.elements.tcp;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+
+import java.security.Principal;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.net.ssl.SSLSession;
+
+import org.eclipse.californium.elements.CorrelationContext;
+import org.eclipse.californium.elements.RawData;
+import org.eclipse.californium.elements.TcpCorrelationContext;
+import org.eclipse.californium.elements.TlsCorrelationContext;
+import org.eclipse.californium.elements.util.StringUtil;
+
+/**
+ * Utils for building for TCP/TLS correlation context and principal from
+ * channel.
+ */
+public class NettyContextUtils {
+
+	private static final Logger LOGGER = Logger.getLogger(NettyContextUtils.class.getName());
+	private static final Level LEVEL = Level.WARNING;
+
+	/**
+	 * Get ssl handler related to the provided channel.
+	 * 
+	 * @param channel channel of ssl handler
+	 * @return ssl handler, or null, if not available.
+	 */
+	private static SslHandler getSslHandler(Channel channel) {
+		ChannelHandler handler = channel.pipeline().first();
+		if (handler instanceof SslHandler) {
+			return (SslHandler) handler;
+		}
+		return null;
+	}
+
+	/**
+	 * Get principal related to the provided channel.
+	 * 
+	 * @param channel channel of principal
+	 * @return principal, or null, if not available
+	 */
+	public static Principal getPrincipal(Channel channel) {
+		SslHandler handler = getSslHandler(channel);
+		if (null != handler) {
+			SSLEngine sslEngine = handler.engine();
+			SSLSession sslSession = sslEngine.getSession();
+			if (null != sslSession) {
+				try {
+					return sslSession.getPeerPrincipal();
+				} catch (SSLPeerUnverifiedException e) {
+					/* ignore it */
+				}
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Build correlation context related to the provided channel.
+	 * 
+	 * @param channel channel of correlation context
+	 * @return correlation context, or null, if yet not available.
+	 */
+	public static CorrelationContext buildCorrelationContext(Channel channel) {
+		String id = channel.id().asShortText();
+		SslHandler handler = getSslHandler(channel);
+		if (null != handler) {
+			SSLEngine sslEngine = handler.engine();
+			SSLSession sslSession = sslEngine.getSession();
+			if (null != sslSession) {
+				byte[] sessionId = sslSession.getId();
+				if (null != sessionId && 0 < sessionId.length) {
```

Comment:
Does someone has experience, if `SSLEngine.getHandshakeStatus()` works better to check, if the session is ready?